src/printer.js

```diff
@@ -2977,6 +2971,50 @@ function hasLeadingOwnLineComment(text, node) {
   return res;
 }
 
+// A "leadingComment" attached to the top of the first token in the
+// return's argument, which would originally be in parens,
+// could trigger an ASI error where the parens are removed
+// and the comment bumped to the same line as the `return`,
+// while the contents of the return value are on the next line (oops!).
+//
+// This recurses the return argument, looking for the first token
+// (the leftmost leaf node) and, if it (or its parents) has any
+// leadingComments, wraps the return in parens.
+function returnArgumentHasLeadingComment(argument) {
+  const hasLeadingComment = (comments) => (
+    comments && comments.some(comment => comment.leading)
+  );
+  const getLeftSide = (node) => {
+    if (node.expressions) return node.expressions[0];
+    return node.left || node.test || node.callee || node.object || node.tag;
+  }
+  const NODE_TYPES_WITH_COMMENTABLE_LEFT_SIDE = {
+    "BinaryExpression": true,
+    "LogicalExpression": true,
+    "ConditionalExpression": true,
+    "CallExpression": true,
+    "MemberExpression": true,
+    "SequenceExpression": true,
+    "TaggedTemplateExpression": true,
+  };
+
+  if (hasLeadingComment(argument.comments)) return true;
+
+  if (NODE_TYPES_WITH_COMMENTABLE_LEFT_SIDE[argument.type]) {
+    let leftMost = argument;
+    let newLeftMost;
+    while (newLeftMost = getLeftSide(leftMost)) {
+      leftMost = newLeftMost;
+
+      if (hasLeadingComment(leftMost.comments)) {
```

Comment:
Note that all comments from the arguments node itself on down ("left" nodes only) are used, which leads to one mildly unfortunate test.

However, several tests from this PR fail without this, and I couldn't think of a clean solution.